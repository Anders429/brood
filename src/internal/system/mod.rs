pub(crate) mod schedule;

mod null;

pub(crate) use null::Null;
