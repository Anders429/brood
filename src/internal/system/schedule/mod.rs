pub(crate) mod raw_task;
pub(crate) mod task;
