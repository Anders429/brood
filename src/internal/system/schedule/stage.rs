use crate::{
    internal::system::schedule::sendable::SendableWorld,
    registry::Registry,
    system::{
        schedule::stage::{Null, Stage},
        ParSystem, System,
    },
};

pub trait StagesSeal<'a>: Send {
    fn run<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn defer<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn run_current<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn run_continuing<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn flush<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;
}

impl<'a> StagesSeal<'a> for Null {
    fn run<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn defer<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn run_current<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn run_continuing<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn flush<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }
}

impl<'a, S, P, L> StagesSeal<'a> for (Stage<S, P>, L)
where
    S: System<'a> + Send,
    P: ParSystem<'a> + Send,
    L: StagesSeal<'a>,
{
    fn run<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        self.defer(world);
        self.run_current(world);
    }

    fn defer<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match self.0 {
            Stage::Start(_) | Stage::Flush => {
                self.1.run(world);
            }
            Stage::Continue(_) => {
                self.1.defer(world);
            }
        }
    }

    fn run_current<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.run(world);
            }
            Stage::Continue(task) => {
                rayon::join(
                    || {
                        self.1.run_continuing(world);
                    },
                    || {
                        task.run(world);
                    },
                );
            }
            Stage::Flush => {
                self.1.flush(world);
            }
        }
    }

    fn run_continuing<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.run(world);
            }
            Stage::Continue(task) => {
                rayon::join(
                    || {
                        self.1.run_continuing(world);
                    },
                    || {
                        task.run(world);
                    },
                );
            }
            Stage::Flush => {}
        }
    }

    fn flush<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.flush(world);
            }
            Stage::Continue(task) => {
                self.1.flush(world);
                task.flush(world);
            }
            Stage::Flush => {}
        }
    }
}
