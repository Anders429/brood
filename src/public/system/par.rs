use crate::{
    query::{filter::Filter, result, view::ParViews},
    registry::Registry,
    world::World,
};

pub trait ParSystem<'a> {
    type Filter: Filter;
    type Views: ParViews<'a> + super::schedule::Claim;

    fn run<R>(&mut self, query_results: result::ParIter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a;

    #[inline]
    #[allow(unused_variables)]
    fn world_post_processing<R>(&mut self, world: &mut World<R>)
    where
        R: Registry,
    {
    }
}
