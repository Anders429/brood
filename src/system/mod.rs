#[cfg(feature = "parallel")]
pub mod schedule;

mod null;
#[cfg(feature = "parallel")]
mod par;

pub use null::Null;
#[cfg(feature = "parallel")]
pub use par::ParSystem;
#[cfg(feature = "parallel")]
pub use schedule::Schedule;

use crate::{
    query::{filter::Filter, result, view::Views},
    registry::Registry,
    world::World,
};

pub trait System<'a> {
    type Filter: Filter;
    type Views: Views<'a>;

    fn run<R>(&mut self, query_results: result::Iter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a;

    #[inline]
    #[allow(unused_variables)]
    fn world_post_processing<R>(&mut self, world: &mut World<R>)
    where
        R: Registry,
    {
    }
}
