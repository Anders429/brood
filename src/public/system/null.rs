#[cfg(feature = "parallel")]
use crate::system::ParSystem;
use crate::{
    query::{filter, result, view},
    registry::Registry,
    system::System,
    world::World,
};
use core::hint::unreachable_unchecked;

pub enum Null {}

impl<'a> System<'a> for Null {
    type Filter = filter::None;
    type Views = view::Null;

    fn run<R>(&mut self, _query_results: result::Iter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a,
    {
        unsafe { unreachable_unchecked() }
    }

    fn world_post_processing<R>(&mut self, _world: &mut World<R>)
    where
        R: Registry,
    {
        unsafe { unreachable_unchecked() }
    }
}

#[cfg(feature = "parallel")]
impl<'a> ParSystem<'a> for Null {
    type Filter = filter::None;
    type Views = view::Null;

    fn run<R>(&mut self, _query_results: result::ParIter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a,
    {
        unsafe { unreachable_unchecked() }
    }

    fn world_post_processing<R>(&mut self, _world: &mut World<R>)
    where
        R: Registry,
    {
        unsafe { unreachable_unchecked() }
    }
}
