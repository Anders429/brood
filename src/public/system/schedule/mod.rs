pub mod stage;
pub mod task;

mod builder;

pub use builder::Builder;

use crate::{
    internal::system::schedule::{sendable::SendableWorld, raw_task}, registry::Registry,
    world::World,
};
use stage::Stages;

pub struct Schedule<S> {
    stages: S,
}

impl Schedule<stage::Null> {
    pub fn builder() -> Builder<raw_task::Null> {
        Builder::new()
    }
}

impl<'a, S> Schedule<S>
where
    S: Stages<'a>,
{
    pub(crate) fn run<R>(&mut self, world: &'a mut World<R>)
    where
        R: Registry,
    {
        self.stages.run(SendableWorld(world));
    }
}
