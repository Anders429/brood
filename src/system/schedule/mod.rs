//! A list of [`System`]s to be run in stages.
//!
//! [`Schedule`]s are created using a builder pattern. `System`s are provided in the desired order
//! they are to be run, and the stages in which those `System`s are run is automatically derived.
//!
//! The advantage of defining a `Schedule` is that `System`s are allowed to be run in parallel as
//! long as their [`Views`] can be borrowed simultaneously.
//!
//! # Example
//! The below example will execute both `SystemA` and `SystemB` in parallel, since their views can
//! be borrowed simultaneously.
//!
//! ``` rust
//! use brood::{
//!     query::{
//!         filter,
//!         filter::Filter,
//!         result,
//!         views,
//!     },
//!     registry::ContainsQuery,
//!     system::{
//!         Schedule,
//!         System,
//!     },
//! };
//!
//! // Define components.
//! struct Foo(usize);
//! struct Bar(bool);
//! struct Baz(f64);
//!
//! struct SystemA;
//!
//! impl<'a> System<'a> for SystemA {
//!     type Views = views!(&'a mut Foo, &'a Bar);
//!     type Filter = filter::None;
//!
//!     fn run<R, FI, VI, P, I, Q>(
//!         &mut self,
//!         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
//!     ) where
//!         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
//!     {
//!         for result!(foo, bar) in query_results {
//!             // Do something...
//!         }
//!     }
//! }
//!
//! struct SystemB;
//!
//! impl<'a> System<'a> for SystemB {
//!     type Views = views!(&'a mut Baz, &'a Bar);
//!     type Filter = filter::None;
//!
//!     fn run<R, FI, VI, P, I, Q>(
//!         &mut self,
//!         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
//!     ) where
//!         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
//!     {
//!         for result!(baz, bar) in query_results {
//!             // Do something...
//!         }
//!     }
//! }
//!
//! let schedule = Schedule::builder().system(SystemA).system(SystemB).build();
//! ```
//!
//! [`Schedule`]: crate::system::schedule::Schedule
//! [`System`]: crate::system::System
//! [`Views`]: crate::query::view::Views

pub mod raw_task;
pub mod stage;

pub(crate) mod task;

mod sendable;

mod builder;

pub use builder::Builder;
pub use stage::stages;

use crate::{
    registry::Registry,
    world::World,
};
use sendable::SendableWorld;
use stage::Stages;

/// A list of [`System`]s to be run in stages.
///
/// The `System`s that make up a `Schedule` are organized into [`Stages`] on creation. `System`s
/// that can be run in parallel are done so. See the documentation for the [`schedule::Builder`]
/// for more information about creating a `Schedule`.
///
/// # Example
/// The below example will execute both `SystemA` and `SystemB` in parallel, since their views can
/// be borrowed simultaneously.
///
/// ``` rust
/// use brood::{
///     query::{
///         filter,
///         filter::Filter,
///         result,
///         views,
///     },
///     registry::ContainsQuery,
///     system::{
///         Schedule,
///         System,
///     },
/// };
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
/// struct Baz(f64);
///
/// struct SystemA;
///
/// impl<'a> System<'a> for SystemA {
///     type Views = views!(&'a mut Foo, &'a Bar);
///     type Filter = filter::None;
///
///     fn run<R, FI, VI, P, I, Q>(
///         &mut self,
///         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
///     ) where
///         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
///     {
///         for result!(foo, bar) in query_results {
///             // Do something...
///         }
///     }
/// }
///
/// struct SystemB;
///
/// impl<'a> System<'a> for SystemB {
///     type Views = views!(&'a mut Baz, &'a Bar);
///     type Filter = filter::None;
///
///     fn run<R, FI, VI, P, I, Q>(
///         &mut self,
///         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
///     ) where
///         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
///     {
///         for result!(baz, bar) in query_results {
///             // Do something...
///         }
///     }
/// }
///
/// let schedule = Schedule::builder().system(SystemA).system(SystemB).build();
/// ```
///
/// [`schedule::Builder`]: crate::system::schedule::Builder
/// [`Stages`]: crate::system::schedule::stage::Stages
/// [`System`]: crate::system::System
#[cfg_attr(doc_cfg, doc(cfg(feature = "rayon")))]
pub struct Schedule<S> {
    stages: S,
}

impl Schedule<stage::Null> {
    /// Creates a [`schedule::Builder`] to construct a new `Schedule`.
    ///
    /// # Example
    /// ``` rust
    /// use brood::system::Schedule;
    ///
    /// let builder = Schedule::builder();
    /// // Add systems to the builder.
    /// let schedule = builder.build();
    /// ```
    ///
    /// [`schedule::Builder`]: crate::system::schedule::Builder
    #[must_use]
    pub fn builder() -> Builder<raw_task::Null> {
        Builder::new()
    }
}

impl<'a, S> Schedule<S> {
    pub(crate) fn run<R, SFI, SVI, PFI, PVI, SP, SI, SQ, PP, PI, PQ>(
        &mut self,
        world: &'a mut World<R>,
    ) where
        R: Registry,
        S: Stages<'a, R, SFI, SVI, PFI, PVI, SP, SI, SQ, PP, PI, PQ>,
    {
        self.stages.run(
            // SAFETY: The pointer provided here is unique, being created from a mutable reference.
            unsafe { SendableWorld::new(world) },
        );
    }
}
