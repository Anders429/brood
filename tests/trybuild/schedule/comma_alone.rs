use brood::system::schedule;

fn main() {
    let schedule = schedule!(,);
}
