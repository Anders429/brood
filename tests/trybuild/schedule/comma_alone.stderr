error: no rules expected the token `,`
 --> tests/trybuild/schedule/comma_alone.rs:4:30
  |
4 |     let schedule = schedule!(,);
  |                              ^ no rules expected this token in macro call
  |
note: while trying to match meta-variable `$task:expr`
 --> src/system/schedule/mod.rs
  |
  |         ($task:expr $(,$tasks:expr)* $(,)?) => (
  |          ^^^^^^^^^^
